### DISCUSSION TITLE
SoftwareSerial on 1606

### DISCUSSION THREAD

**Discussion by raelectronics** [SpenceKonde/megaTinyCore#653] [Library discussion]
First of all, thank you for your work!
I try to receive data using SoftwareSerial on a 1606 uC (PA5 for TX & PA6 for RX), but after a few seconds of operation (2 to 5 sec) it freezes. In fact, I only need RX (1200 baud-rate) to receive the data sent by an old device (8N2). I have been trying to solve this problem for a few days and a piece of advice would help me a lot.
Thank you!

**Comment by SpenceKonde** [comment]
set attachInterruptMode to Old Version. 

SoftwareSerial was the library that I most wanted to address (the whole motivation was the way that attackInterurrupt grabs every single port vector. And since attach'ed interrupts are SLOW AS ALL HELL that means you can't use any library that used attachInterrupt in combination with a port interrupt that got reasonable performance. And unlike on classic AVRs, where there were the PCINTs (which SoftwareSerial grabbed every one of) and the INTn interrupts, here they just gave PCINTs all the features of the INTn interrupots and more, which is great, except that now attachInterrupt() leaves you with NO WAY to interrupt on a pin except attachInterrupt., My new attachInterrupt implementation was aimed at providing a way to choose not to enable attachInterrupt on all ports. (by default they are all enabled though, for compatibility).

But it doesn't work and  I do not understand what is going wrong. The new implementation is a collection of extremely weird assembly to prevent duplication of code in the compiled binary... but most of the techniques involved are used in the Serial implementation too, and they work there. And of course I dont have code that I know reproduces it consistently (It worked fine in my tests!) so I can't even be sure that any fix worked. So far all the attachInterruipt bug manifesting code has been stuff that used external events, so it's useless for reproducing the bug without a way to also mimic the input it's getting. 

I can't work on this until ATTinyCore 2.0.0 is released.

**Comment by raelectronics** [comment]
Thank you for your answer.
I modified the library, implementing your example of using interrupts, specifically for PORTA.PIN6 (D2) and it works perfectly. Once again, thank you for your response and for your work.

**Comment by SpenceKonde** [comment]
I think I have a fix in for the Software Serial library. in WInterrupts.c - though manually defined interrupts are far better performing  than attached ones >.>
